dialer.go

```diff
@@ -0,0 +1,188 @@
+	Label() string
+}
+
+// Stats is also provided by the multipath dialer so the caller can get and
+// print the status of each path.
+type Stats interface {
+	FormatStats() (stats []string)
+}
+
+type subflowDialer struct {
+	Dialer
+	label            string
+	successes        uint64
+	consecSuccesses  uint64
+	failures         uint64
+	framesSent       uint64
+	framesRetransmit uint64
+	framesRecv       uint64
+	bytesSent        uint64
+	bytesRetransmit  uint64
+	bytesRecv        uint64
+	emaRTT           *ema.EMA
+}
+
+func (sfd *subflowDialer) DialContext(ctx context.Context) (net.Conn, error) {
+	conn, err := sfd.Dialer.DialContext(ctx)
+	if err == nil {
+		atomic.AddUint64(&sfd.successes, 1)
+		atomic.AddUint64(&sfd.consecSuccesses, 1)
+	} else {
+		// reset RTT to deprioritize this dialer
+		sfd.emaRTT.SetDuration(longRTT)
+		atomic.AddUint64(&sfd.failures, 1)
+		atomic.StoreUint64(&sfd.consecSuccesses, 0)
+	}
+	return conn, err
+}
+
+func (sfd *subflowDialer) OnRecv(n uint64) {
+	atomic.AddUint64(&sfd.framesRecv, 1)
+	atomic.AddUint64(&sfd.bytesRecv, n)
+}
+func (sfd *subflowDialer) OnSent(n uint64) {
+	atomic.AddUint64(&sfd.framesSent, 1)
+	atomic.AddUint64(&sfd.bytesSent, n)
+}
+func (sfd *subflowDialer) OnRetransmit(n uint64) {
+	atomic.AddUint64(&sfd.framesRetransmit, 1)
+	atomic.AddUint64(&sfd.bytesRetransmit, n)
+}
+func (sfd *subflowDialer) UpdateRTT(rtt time.Duration) {
+	sfd.emaRTT.UpdateDuration(rtt)
+}
+
+type mpDialer struct {
+	name    string
+	dialers []*subflowDialer
+}
+
+func MPDialer(name string, dialers []Dialer) Dialer {
+	var subflowDialers []*subflowDialer
+	for _, d := range dialers {
+		subflowDialers = append(subflowDialers, &subflowDialer{Dialer: d, label: d.Label(), emaRTT: ema.NewDuration(longRTT, rttAlpha)})
+	}
+	d := &mpDialer{name, subflowDialers}
+	return d
+}
+
+// DialContext dials the addr using all dialers and returns a connection
+// contains subflows from whatever dialers available.
+func (mpd *mpDialer) DialContext(ctx context.Context) (net.Conn, error) {
+	var bc *mpConn
+	dialOne := func(d *subflowDialer, cid connectionID) (connectionID, bool) {
+		conn, err := d.DialContext(ctx)
+		if err != nil {
+			log.Errorf("failed to dial %s: %v", d.Label(), err)
+			return zeroCID, false
+		}
+		probeStart := time.Now()
+		newCID, err := mpd.handshake(conn, cid)
+		if err != nil {
+			log.Errorf("failed to handshake %s, continuing: %v", d.Label(), err)
+			conn.Close()
+			return zeroCID, false
+		}
+		if cid == zeroCID {
+			bc = newMPConn(newCID)
+		}
+		bc.add(d.label, conn, true, probeStart, d)
+		return newCID, true
+	}
+	dialers := mpd.sorted()
+	for i, d := range dialers {
+		// dial the first connection with zero connection ID
+		cid, ok := dialOne(d, zeroCID)
+		if !ok {
+			continue
+		}
+		if i < len(dialers)-1 {
+			// dial the rest in parallel with server assigned connection ID
+			for _, d := range dialers[i+1:] {
+				go dialOne(d, cid)
+			}
+		}
+		return bc, nil
+	}
```

Comment:
There could be several strategies like paced dialing or dialing on demand but I don't feel strong on needing them. Let's KISS.